glean-core/csharp/Glean/Dispatchers.cs

```diff
@@ -39,61 +39,44 @@ internal class Dispatchers
         private static int _queueInitialTasks = 1;
         /// <summary>
         /// When true, tasks will be queued and not ran until triggered by
-        /// calling FlushQueuedInitialTasks.
+        /// calling FlushQueuedInitialTasks.  This uses an int backing field
+        /// in order to take advantage of Interlocked.Exchange for thread
+        /// safety.
         /// </summary>
         internal static bool QueueInitialTasks
         {
             get => _queueInitialTasks == 1;
             set => Interlocked.Exchange(ref _queueInitialTasks, value ? 1 : 0);
         }
 
-        // Create a ConcurrentExclusiveSchedulerPair object.
-        // All API tasks will be executed on the exclusive part of the scheduler
-        // in order to ensure single threaded behavior
-        private static readonly
-            ConcurrentExclusiveSchedulerPair taskSchedulerPair =
-            new ConcurrentExclusiveSchedulerPair();
-
-        /// <summary>
-        /// This Queue holds the initial tasks that are launched before Glean is
-        /// initialized.
-        /// </summary>
-        internal static ConcurrentQueue<Action> taskQueue =
-            new ConcurrentQueue<Action>();
+        // Create a scheduler that uses a single thread
+        private static readonly LimitedConcurrencyLevelTaskScheduler lcts =
+            new LimitedConcurrencyLevelTaskScheduler(1);
 
-        /// <summary>
-        /// The number of tasks added to the queue beyond the MaxQueueSize.
-        /// </summary>
-        private static int OverflowCount { get; set; } = 0;
+        // Create a new TaskFactory and pass it the scheduler and cancellation
+        // token.
+        private static readonly TaskFactory factory = new TaskFactory(lcts);
 
         /// <summary>
-        /// Provides the token source for cancellation tokens used to cancel
-        /// running background tasks.
+        /// This Queue holds the initial Actions that are launched before Glean
+        /// is initialized.
         /// </summary>
-        private static readonly CancellationTokenSource tokenSource =
-            new CancellationTokenSource();
+        internal static List<Action> preInitActionQueue = new List<Action>();
```

Comment:
HA! Now I see what the problem was (now that you fixed it somewhere else in the code :) ). This should really be a [`ConcurrentQueue`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentqueue-1?view=netcore-3.1)! We use a `ConcurrentLinkQueue` in Kotlin (lock-free), but it's fine to use any other thing here as log as it is thread safe (and `ConcurrentQueue` is).

Reply:
I can change it back now that I have fixed the real issue :)